docs/rfcs/0003-rfc-controller_refactor.md

```diff
@@ -0,0 +1,99 @@
+- Status: In review
+- Date: 2023-06-20
+- Author: [@gela](https://github.com/gela)
+- RFC PR: https://github.com/mozilla-mobile/mozilla-vpn-client/pull/7269
+- Implementation GitHub issue: TBD.
+
+
+## Summary
+Introduce a new `ConnectionManager` component which sits between the frontend and the controller to handle all connectivity related logic and keep the controller code lean and minimal.
+
+## Motivation
+The current controller code is complicated, lengthy, and bug-prone. Over time we have continued to directly add more logic and checks for various scenarios to the controller code, resulting in a Frankenstein monster that is hard to scale and keep under control.
+
+Because our current approach is not scalable, we’ve been facing numerous issues in the connectivity space, especially when it comes down to identifying exactly what is causing the issue. This means the behavior of our client is extremely inconsistent both on different platforms but also within the same environment. For example, if a user loses network connectivity while the VPN is activated, sometimes they enter no signal, and other times they encounter the “server location unavailable” modal. This inconsistency is a result of racing checks and error reporting. The goal of this refactor is to combine all of these scenarios under one component and to get rid of the `ConnectionHealth` entirely to resolve this issue.
+
+
+## Proposed Solution
+Move all connectivity code out of the controller and into its separate component which is referred to as the `ConnectionManager` in this document. The controller object will continue to remain but its purpose will be drastically simplified to a few basic functionalities including `activate()` and `deactivate()`. We will no longer have any states associated with the Controller and all states will live in the `ConnectionManager`. The `ConnectionManager` will primarily act as a liaison between the controller and the UI. 
```

Comment:
nit:
```suggestion
Move all connectivity code out of the controller and into its separate component which is referred to as the `ConnectionManager` in this document. 

The controller object will remain, but its purpose will be drastically simplified to a few basic functionalities including `activate()` and `deactivate()`. Ther will no longer be any states associated with the Controller and all states will live in the `ConnectionManager`. The `ConnectionManager` will primarily act as a liaison between the controller and the UI. 
```

Reply:
Would you mind expanding here on what are the advantages of moving state management to the ConnectionManager?